### PR TITLE
Add OS/architecture classifier to distributions

### PR DESCRIPTION
**Pull request by rjernst**
This commit adds classifiers to the distributions indicating the
OS (for archives) and platform. The current OSes are for windows, darwin (ie
macos) and linux. This change will allow future OS/architecture specific
changes to the distributions. Note the docs using distribution links
have been updated, but will be reworked in a followup to make OS
specific instructions for the archives.

### CODE REVIEW

**Comment by rjernst** [comment]
@elasticmachine test this please

**Comment by rjernst** [comment]
@elasticmachine test this please

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1

File: distribution/bwc/build.gradle [hunk 1]
```diff
@@ -216,9 +233,15 @@ bwcVersions.forPreviousUnreleased { VersionCollection.UnreleasedVersionInfo unre
         String artifactFileName = artifactFile.name
         String artifactName = artifactFileName.contains('oss') ? 'elasticsearch-oss' : 'elasticsearch'
         String suffix = artifactFile.toString()[-3..-1]
+        int archNdx = artifactFileName.indexOf('x86_64')
```
Comment (jasontedor) [inline distribution/bwc/build.gradle]: Maybe `Index` instead of `Ndx` please?

File: distribution/bwc/build.gradle [hunk 2]
```diff
@@ -122,17 +122,34 @@ bwcVersions.forPreviousUnreleased { VersionCollection.UnreleasedVersionInfo unre
...
+    } else {
+      projects.add('zip')
+    }
+
     for (String projectName : projects) {
         String baseDir = "distribution"
+        String classifier = ""
+        String extension = projectName
+        if (bwcVersion.onOrAfter('7.0.0') && (projectName.contains('zip') || projectName.contains('tar'))) {
+          int ndx = projectName.indexOf('-')
```
Comment (jasontedor) [inline distribution/bwc/build.gradle]: `index` please. 🙏

File: distribution/bwc/build.gradle [hunk 3]
```diff
@@ -216,9 +233,15 @@ bwcVersions.forPreviousUnreleased { VersionCollection.UnreleasedVersionInfo unre
         String artifactFileName = artifactFile.name
         String artifactName = artifactFileName.contains('oss') ? 'elasticsearch-oss' : 'elasticsearch'
         String suffix = artifactFile.toString()[-3..-1]
+        int archNdx = artifactFileName.indexOf('x86_64')
+        String classifier = ''
+        if (archNdx != -1) {
+            int osNdx = artifactFileName.lastIndexOf('-', archNdx - 2)
```
Comment (jasontedor) [inline distribution/bwc/build.gradle]: `Index` instead of `Ndx`.

**Review by jasontedor — APPROVED** [review]
Looks good. Just one request on not abbreviating some variable names; no need for another round.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/default-distro

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/default-distro

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/default-distro

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**